pkg/auth/dependency/authn/oauth.go

```diff
@@ -14,32 +14,38 @@ type OAuthCoordinator struct {
 	Signup *SignupProcess
 }
 
-func (c *OAuthCoordinator) AuthenticateCode(authInfo sso.AuthInfo, codeChallenge string, loginState sso.LoginState) (code *sso.SkygearAuthorizationCode, err error) {
+func (c *OAuthCoordinator) AuthenticateCode(authInfo sso.AuthInfo, codeChallenge string, loginState sso.LoginState) (*sso.SkygearAuthorizationCode, string, error) {
```

Comment:
It is more intuitive that `AuthenticateCode` and `LinkCode` also persist the code before returning. So `SSOProvider.StoreSkygearAuthorizationCode` can be removed.

In some flows, the code is consumed immediately. We can actually consume the code.

Reply:
If the code is persisted in `OAuthCoordinator`, it should be consume there too. Moved both store and consume to `OAuthCoordinator` in 61f79eb. See if it looks good?

Reply:
@kiootic Which behavior do you prefer? As you have mentioned in your PR that authenticate should not have side effects...

Sorry Carmen we may want to revert that change. :(

Reply:
I think consuming the authorization code is okay, since SSO authentication is performed (i.e. code is consumed) before entering the new interaction flow. The new interaction flow should not have externally visible side effect (e.g. consumption of code) before committing the interaction.